visionmetrics/detection/confusion_matrix.py

```diff
@@ -0,0 +1,171 @@
+
+            # Check empty predictions and targets
+            if self._is_empty(preds):
+                self.fn += len(gts)
+                continue
+
+            if self._is_empty(gts):
+                self.fp += len(preds)
+                self.fp_due_to_extra_pred_boxes += len(preds)
+                continue
+
+            # Count per image
+            gt_boxes_used = [False] * len(gts)  # To keep track of all matched GT boxes
+            preds = sorted(preds, key=lambda x: x[1], reverse=True)
+            for pred in preds:
+                # If all GT boxes for this image have been matched (TP)
+                # with corresponding pred boxes then all remaining pred boxes are FPs
+                if all(gt_boxes_used):
+                    self.fp += 1
+                    self.fp_due_to_extra_pred_boxes += 1
+                    continue
+
+                # Otherwise, calculate IoU with all remaining GT boxes
+                pred_class_id, pred_box = pred[0], pred[2:]
+                best_same_class_iou = 0
+                best_same_class_gt_index = -1
+                best_diff_class_iou = 0
+
+                for gt_index, gt in enumerate(gts):
+                    if gt_boxes_used[gt_index]:  # Skip GT box if already used
+                        continue
+
+                    gt_class_id, gt_box = gt[0], gt[1:]
+
+                    # Calculate IoU and update best pred box IoU with same and different class GT boxes
+                    current_iou = self.iou(pred_box, gt_box)
+                    if pred_class_id == gt_class_id:
+                        if current_iou > best_same_class_iou:
+                            best_same_class_iou = current_iou
+                            best_same_class_gt_index = gt_index
+                    else:
+                        if current_iou > best_diff_class_iou:
+                            best_diff_class_iou = current_iou
+
+                # Update TP and FP counts
+                if best_same_class_gt_index != -1:  # GT box with same class found
+                    if best_same_class_iou >= iou_threshold:
+                        self.tp += 1
+                        gt_boxes_used[best_same_class_gt_index] = True  # Mark GT box as used
+                    else:
+                        self.fp += 1
+                        self.fp_due_to_low_iou += 1
+                else:
+                    if best_diff_class_iou >= iou_threshold:
+                        self.fp += 1
+                        self.fp_due_to_wrong_class += 1
+                    else:
+                        self.fp += 1
+                        self.fp_due_to_low_iou += 1
+
+            # Count unused GT boxes as FNs
+            for gt_index, gt in enumerate(gts):
+                if not gt_boxes_used[gt_index]:
+                    self.fn += 1
+
+    @staticmethod
+    def _is_empty(pred_or_gt):
+        return all([len(x) == 0 for x in pred_or_gt])
+
+    @staticmethod
```

Comment:
torchmetrics*IOU excepts a list of dicts comprising tensors. Our preds and targets are nested lists of floats. So, I preferred to write a simple method to calculate IOU instead of doing float -> tensor conversions and back.